packages/core/src/zip/selector.js

```diff
@@ -0,0 +1,96 @@
+'use strict'
+
+import R from 'ramda'
+
+import * as zip from './impl'
+
+import { isOfKind } from '../data'
+
+import { isStringArray } from './select'
+
+// Selector fns
+// - A selector fn has the signature: (...args, loc) => [loc]
+// - It takes a number of arguments that determine its behavior
+// - a zipper location at the end
+// - and will return a an array of locations that have been selected by its execution
+
+const isNotChildCollection = R.complement(isOfKind('@@skele/child-collection'))
+
+// children :: Location -> [Location]
+export const elementChildren = loc => elementChildrenFor(null, loc)
+
+// Key = String | [String]
+// childrenFor :: Key -> Location -> [Location]
+export const elementChildrenFor = R.curry((key, loc) => {
+  const result = []
+  let childLoc = zip.down(loc)
+  if (childLoc != null) {
+    const child = zip.node(childLoc)
+    if (
+      R.isNil(key) ||
+      (isStringArray(key) && R.contains(child.get('propertyName'), key)) ||
+      (R.is(String)(key) && child.get('propertyName') === key)
+    ) {
+      if (child.get('isSingle')) {
+        result.push(zip.down(childLoc))
+      } else {
+        let currentChild = zip.down(childLoc)
+        result.push(currentChild)
+        while ((currentChild = zip.right(currentChild)) != null) {
+          result.push(currentChild)
+        }
+      }
+    }
+    while ((childLoc = zip.right(childLoc)) != null) {
+      const child = zip.node(childLoc)
+      if (
+        R.isNil(key) ||
+        (isStringArray(key) && R.contains(child.get('propertyName'), key)) ||
+        (R.is(String)(key) && child.get('propertyName') === key)
+      ) {
+        if (child.get('isSingle')) {
+          result.push(zip.down(childLoc))
+        } else {
+          let currentChild = zip.down(childLoc)
+          result.push(currentChild)
+          while ((currentChild = zip.right(currentChild)) != null) {
+            result.push(currentChild)
+          }
+        }
+      }
+    }
+  }
+  return result
+})
+
+// ancestors :: () -> Location -> [Location]
+export const ancestors = loc => {
+  const result = []
+  let currentLoc = loc
+  while ((currentLoc = zip.up(currentLoc)) != null) {
+    if (isNotChildCollection(zip.node(currentLoc))) {
+      result.push(currentLoc)
+    }
+  }
+  return result
+}
+
+// descendants :: () -> Location -> [Location]
+export const descendants = loc => _descendants(loc)
+
+const _descendants = (loc, collector = []) => {
```

Comment:
I'm having fears about this being an eager implementation. 

What if a) our zipper implementation and select does not treat children as a concrete array but an 'iterable' and
b) we implement this as a LazySeq? Is that doable?

Reply:
I think this is doable. Here is how:
```diff
// descendants :: Location -> [Location]
- export const descendants = loc => _descendants(loc)
+ export const descendants = loc => [..._descendants(loc)]

- const _descendants = (loc, collector = []) => {
+ const _descendants = function*(loc) {
  let current = null
  if ((current = zip.down(loc)) != null) {
-    _descendants(current, collector)
+    yield* _descendants(current)
-    collector.push(current)
+    yield current
  }
  if ((current = zip.right(loc)) != null) {
-    _descendants(current, collector)
+    yield* _descendants(current)
-    collector.push(current)
+    yield current
  }
-  return collector
}
```
Of course, here I use the generator function with the spread syntax, so that I keep the same interface to check that tests are giving same results.

To make it lazy-sequence I would break/change the interface for "selectors", and it will look like this:
```diff
+import I from 'immutable'
- // descendants :: Location -> [Location]
+ // descendants :: Location -> Seq(...locations)
- export const descendants = loc => [..._descendants(loc)]
+ export const descendants = loc => I.Seq(_descendants(loc))
```

What do you think?

Reply:
Implemented in https://github.com/netceteragroup/skele/pull/113/commits/92c7ce03316e67252e7673940fd068de12f241e3